sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownUtils.scala

```diff
@@ -76,28 +78,46 @@ object PushDownUtils extends PredicateHelper {
    * @return the created `ScanConfig`(since column pruning is the last step of operator pushdown),
    *         and new output attributes after column pruning.
    */
-  // TODO: nested column pruning.
   def pruneColumns(
       scanBuilder: ScanBuilder,
       relation: DataSourceV2Relation,
-      exprs: Seq[Expression]): (Scan, Seq[AttributeReference]) = {
+      projects: Seq[NamedExpression],
+      filters: Seq[Expression]): (Scan, Seq[AttributeReference]) = {
     scanBuilder match {
+      case r: SupportsPushDownRequiredColumns if SQLConf.get.nestedSchemaPruningEnabled =>
```

Comment:
If we know scanBuilder does not support nested schema pruning, do we still need go this path?

Reply:
I don't think there is an API for find out whether a particular `ScanBuilder` supports nested schema pruning or not. Instead, we have only `SupportsPushDownRequiredColumns` and data sources should use the passed schema as a reference and prune whatever they can. The flag that we added in this PR is specific to `FileScanBuilder` and I believe it will complicate the overall logic if we treat `FileScanBuilder` differently by introducing special branches or if conditions. Also, we might actually get rid of that flag soon, as mentioned in DB's [comment](https://github.com/apache/spark/pull/26751#discussion_r356771927).